x-pack/plugins/maps/public/actions/visualize_geo_field_action.ts

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import rison from 'rison-node';
+import uuid from 'uuid/v4';
+import { stringify } from 'query-string';
+import { i18n } from '@kbn/i18n';
+import { createAction } from '../../../../../src/plugins/ui_actions/public';
+import {
+  createKbnUrlStateStorage,
+  IKbnUrlStateStorage,
+} from '../../../../../src/plugins/kibana_utils/public';
+import { DiscoverAppState } from '../../../../../src/plugins/discover/public';
+import { getApplication, getIndexPatternService } from '../kibana_services';
+
+export const ACTION_VISUALIZE_GEO_FIELD = 'ACTION_VISUALIZE_GEO_FIELD';
+
+export const visualizeGeoFieldAction = createAction<typeof ACTION_VISUALIZE_GEO_FIELD>({
+  type: ACTION_VISUALIZE_GEO_FIELD,
+  getDisplayName: () =>
+    i18n.translate('xpack.maps.discover.visualizeFieldLabel', {
+      defaultMessage: 'Visualize on Kibana Maps',
+    }),
+  execute: async (context) => {
+    const indexPattern = await getIndexPatternService().get(context.indexPatternId);
+    const field = indexPattern.fields.find((fld) => fld.name === context.fieldName);
+    const supportsClustering = field?.aggregatable;
+    const stateStorage: IKbnUrlStateStorage = createKbnUrlStateStorage();
+    const appStateFromUrl: DiscoverAppState | null = stateStorage.get('_a');
+    // create initial layer descriptor
+    const hasTooltips =
+      appStateFromUrl?.columns?.length && appStateFromUrl?.columns[0] !== '_source';
+
+    const linkUrlParams = {
```

Comment:
Suggest to abstract url building to urlGenerators. 

We already have url generators for discover, dashboard and ml app. 
In this case worth creating visualize app and map app url generators. 

* docs: https://github.com/elastic/kibana/blob/master/src/plugins/share/public/url_generators/README.md
* there are also example plugin: https://github.com/elastic/kibana/blob/master/examples/url_generators_explorer/README.md
* Example of dashboard url generator: 
  * https://github.com/elastic/kibana/blob/master/src/plugins/dashboard/public/url_generator.ts (Also notice `setStateToKbnUrl` instead direct rison usage)
  * And usage: https://github.com/elastic/kibana/blob/master/x-pack/plugins/dashboard_enhanced/public/services/drilldowns/dashboard_to_dashboard_drilldown/drilldown.tsx#L117 from dashboard drilldown

Reply:
The benefit on that approach is that apps own there urls, so apps provide services for other apps to generate URL to deep link into them. 

We recommend to so app exposes the urlGenerator on its plugin contact, but if direct plugin decency is not possible or doesn't make sense, then it could be used as registry

Reply:
@Dosant I have created url generators for maps and visualize as you proposed. Thanx for pointing this out :)